### PR TITLE
feat: Add `col.types` argument to `duckdb_read_csv()`

### LINKED ISSUE #383: R: Add support for the `type`/`dtypes` in `duckdb_read_csv`?
I am trying to read the [Fannie Mae data](https://capitalmarkets.fanniemae.com/credit-risk-transfer/single-family-credit-risk-transfer/fannie-mae-single-family-loan-performance-data) which is a darling of the larger than RAM world.

But I can't seem to read it into R. I checked the source code and it doesn't support types yet.

For example, in disk.frame, we can read them like this by specifying the types but duckdb can't at the moment.

```
library(disk.frame) 
setup_disk.frame() # limit to two worker so as not to blow the memory usage

#library(future)

#future::plan(sequential, split = TRUE)

lppub_column_names <- c("POOL_ID", "LOAN_ID", "ACT_PERIOD", "CHANNEL", "SELLER", "SERVICER",
                        "MASTER_SERVICER", "ORIG_RATE", "CURR_RATE", "ORIG_UPB", "ISSUANCE_UPB",
                        "CURRENT_UPB", "ORIG_TERM", "ORIG_DATE", "FIRST_PAY", "LOAN_AGE",
                        "REM_MONTHS", "ADJ_REM_MONTHS", "MATR_DT", "OLTV", "OCLTV",
                        "NUM_BO", "DTI", "CSCORE_B", "CSCORE_C", "FIRST_FLAG", "PURPOSE",
                        "PROP", "NO_UNITS", "OCC_STAT", "STATE", "MSA", "ZIP", "MI_PCT",
                        "PRODUCT", "PPMT_FLG", "IO", "FIRST_PAY_IO", "MNTHS_TO_AMTZ_IO",
                        "DLQ_STATUS", "PMT_HISTORY", "MOD_FLAG", "MI_CANCEL_FLAG", "Zero_Bal_Code",
                        "ZB_DTE", "LAST_UPB", "RPRCH_DTE", "CURR_SCHD_PRNCPL", "TOT_SCHD_PRNCPL",
                        "UNSCHD_PRNCPL_CURR", "LAST_PAID_INSTALLMENT_DATE", "FORECLOSURE_DATE",
                        "DISPOSITION_DATE", "FORECLOSURE_COSTS", "PROPERTY_PRESERVATION_AND_REPAIR_COSTS",
                        "ASSET_RECOVERY_COSTS", "MISCELLANEOUS_HOLDING_EXPENSES_AND_CREDITS",
                        "ASSOCIATED_TAXES_FOR_HOLDING_PROPERTY", "NET_SALES_PROCEEDS",
                        "CREDIT_ENHANCEMENT_PROCEEDS", "REPURCHASES_MAKE_WHOLE_PROCEEDS",
                        "OTHER_FORECLOSURE_PROCEEDS", "NON_INTEREST_BEARING_UPB", "PRINCIPAL_FORGIVENESS_AMOUNT",
                        "ORIGINAL_LIST_START_DATE", "ORIGINAL_LIST_PRICE", "CURRENT_LIST_START_DATE",
                        "CURRENT_LIST_PRICE", "ISSUE_SCOREB", "ISSUE_SCOREC", "CURR_SCOREB",
                        "CURR_SCOREC", "MI_TYPE", "SERV_IND", "CURRENT_PERIOD_MODIFICATION_LOSS_AMOUNT",
                        "CUMULATIVE_MODIFICATION_LOSS_AMOUNT", "CURRENT_PERIOD_CREDIT_EVENT_NET_GAIN_OR_LOSS",
                        "CUMULATIVE_CREDIT_EVENT_NET_GAIN_OR_LOSS", "HOMEREADY_PROGRAM_INDICATOR",
                        "FORECLOSURE_PRINCIPAL_WRITE_OFF_AMOUNT", "RELOCATION_MORTGAGE_INDICATOR",
                        "ZERO_BALANCE_CODE_CHANGE_DATE", "LOAN_HOLDBACK_INDICATOR", "LOAN_HOLDBACK_EFFECTIVE_DATE",
                        "DELINQUENT_ACCRUED_INTEREST", "PROPERTY_INSPECTION_WAIVER_INDICATOR",
                        "HIGH_BALANCE_LOAN_INDICATOR", "ARM_5_YR_INDICATOR", "ARM_PRODUCT_TYPE",
                        "MONTHS_UNTIL_FIRST_PAYMENT_RESET", "MONTHS_BETWEEN_SUBSEQUENT_PAYMENT_RESET",
                        "INTEREST_RATE_CHANGE_DATE", "PAYMENT_CHANGE_DATE", "ARM_INDEX",
                        "ARM_CAP_STRUCTURE", "INITIAL_INTEREST_RATE_CAP", "PERIODIC_INTEREST_RATE_CAP",
                        "LIFETIME_INTEREST_RATE_CAP", "MARGIN", "BALLOON_INDICATOR",
                        "PLAN_NUMBER", "FORBEARANCE_INDICATOR", "HIGH_LOAN_TO_VALUE_HLTV_REFINANCE_OPTION_INDICATOR",
                        "DEAL_NAME", "RE_PROCS_FLAG", "ADR_TYPE", "ADR_COUNT", "ADR_UPB", "PAYMENT_DEFERRAL_MOD_EVENT_FLAG", "INTEREST_BEARING_UPB")

lppub_column_names = tolower(lppub_column_names)

lppub_column_classes <- c("character", "character", "character", "character", "character", "character",
                          "character", "numeric", "numeric", "numeric", "numeric",
                          "numeric", "numeric", "character", "character", "numeric", "numeric",
                          "numeric", "character", "numeric", "numeric", "character", "numeric",
                          "numeric", "numeric", "character", "character", "character",
                          "numeric", "character", "character", "character", "character",
                          "numeric", "character", "character", "character", "character",
                          "numeric", "character", "character", "character", "character",
                          "character", "character", "numeric", "character", "numeric",
                          "numeric", "numeric", "character", "character", "character",
                          "numeric", "numeric", "numeric", "numeric", "numeric", "numeric",
                          "numeric", "numeric", "numeric", "numeric", "numeric", "character",
                          "numeric", "character", "numeric", "numeric", "numeric", "numeric",
                          "numeric", "numeric", "character", "numeric", "numeric", "numeric",
                          "numeric", "character", "numeric", "character", "numeric", "character",
                          "numeric", "numeric", "character", "character", "numeric", "numeric",
                          "numeric", "numeric", "numeric", "numeric", "numeric", "numeric",
                          "numeric", "numeric", "numeric", "numeric", "numeric", "character",
                          "character", "character", "character", "character",
                          "character", "numeric", "numeric", "character", "numeric")


source("scrap.R")

# the data contained all the data from 
system.time(fm.df <- zip_to_disk.frame(
  "c:/data/fannie-mae-single-family/Performance_All.zip",
  outdir = "c:/data/fannie-mae-disk.frame/", 
  # passed to fread
  col.names = lppub_column_names,
  colClasses = lppub_column_classes,
  shardby = "loan_id"
  ))
                  

system.time(a <- list.dirs("c:/data/fannie-mae-disk.frame/", recursive = F) %>% 
  lapply(disk.frame) %>% 
  rbindlist.disk.frame(outdir = "c:/data/fannie-mae-disk.frame2/"))


a = disk.frame("c:/data/fannie-mae-disk.frame2/")
head(a)

system.time(b <- a %>% 
              group_by(pool_id) %>% 
              summarise(n()) %>% 
              collect)
```

### PR DESCRIPTION
**Pull request by eli-daniels**
This addresses 2 issues: https://github.com/duckdb/duckdb-r/issues/118 and https://github.com/duckdb/duckdb-r/issues/383

col.types can be given as a named character vector providing the names and types of the columns such as: `col.types = c(col0 = 'VARCHAR', col1 = 'COUBLE' , etc..)` or an unnamed character vector, then col names are taken from the read.csv output or the col.names argument. Column names given by col.types are preferred over col.names.

The data types are provided as DuckDB data types, so VARCHAR, DOUBLE, BIGINT, etc...

As part of this I changed dbWriteTable to dbCreateTable. So could easily add the temporary parameter mentioned here https://github.com/duckdb/duckdb-r/issues/142, but don't want to get ahead of myself.

Also made a minor addition to the docs to mention that the csv files are appended to the table if the table already exists, I hope that is okay.

Happy to modify anything that may be amiss! 

Closes #383.

### CODE REVIEW

**Comment by krlmlr** [comment]
Thanks for the effort, good idea!

This function is getting too complex now, is there a way to split this into more focused functions, one of which would be responsible for inferring the column types? Can you propose a refactoring that will later make it easy to add this functionality?

The branch now also conflicts with the main branch, sorry about that.

File: R/csv.R [hunk 1]
```diff
@@ -12,6 +14,9 @@
 #' @param delim Which field separator should be used
 #' @param quote Which quote character is used for columns in the CSV file
 #' @param col.names Override the detected or generated column names
+#' @param col.types Character vector of column types in the same order as col.names,
+#' or a named character vector where names are column names and types pairs.
+#' Valid ypes are DuckDB data types, e.g. VARCHAR, DOUBLE, DATE, BIGINT, BOOLEAN, etc.
```
Comment (xiaodaigh) [inline R/csv.R:19]: instead of saying etc, can we point to a documentation somewhere?

**Comment by eli-daniels** [comment]
Yep, I'll give it a shot and add some tests and better docs.

Is there any way to avoid recompiling DuckDB when doing RCMD checks?

**Comment by krlmlr** [comment]
I use `ccache` to speed up recompilation: https://stackoverflow.com/a/45512708/946850, https://dirk.eddelbuettel.com/blog/2017/11/27/ .

**Comment by eli-daniels** [comment]
I've refactored it a bit and added a DATE type test. Ready for review.

If there is anything else I can do to make it better, let me know.

RCMD Check gives : [ FAIL 0 | WARN 0 | SKIP 52 | PASS 4810 ]

File: R/csv.R [hunk 2]
```diff
@@ -100,6 +119,55 @@ duckdb_read_csv <- function(
   invisible(out)
 }
 
+
+#' Column names and types logic for duckdb_read_csv()
+#'
+#' @param found the detected (found) header and types from `utils::read_csv`
+#' @param col.names user provided column names
+#' @param col.types user provider column types and maybe names too
+#'
```
Comment (krlmlr) [inline R/csv.R:128]:
```suggestion
#'
#' @noRd
```

To avoid generation of `.Rd` file.

File: R/csv.R [hunk 3]
```diff
@@ -100,6 +119,55 @@ duckdb_read_csv <- function(
...
+      return(col.types)
+    } else {
+      if (length(col.types) != ncol(found)) {
+        stop(
+          "You supplied ", length(col.types), " values to `col.types`, but file has ",
+          ncol(found), " columns."
+        )
+      }
+      fields <- col.types
+      name(fields) <- col.names
```
Comment (krlmlr) [inline R/csv.R:158]: `R CMD check` complains, maybe:

```suggestion
      names(fields) <- col.names
```

**Review by krlmlr — COMMENTED** [review]
Thanks, looks like CI/CD is failing.

**Comment by eli-daniels** [comment]
Thanks for the review @krlmlr

**Comment by krlmlr** [comment]
Thanks! This is a good start, but ideally, I'd like to proceed as outlined in https://github.com/duckdb/duckdb-r/issues/118#issuecomment-2418569594 . Merging for now, let's see.

**Comment by eli-daniels** [comment]
Agreed, but this adds the functionality asked across a few issues. Hopefully it helps someone. I'm still thinking about how to approach the method discussed in #118

**Pull request merged**